plugins/metrics/pkg/gateway/services/rbac.go

```diff
@@ -0,0 +1,257 @@
+		Source:           "plugin_metrics",
+		AvailableDrivers: drivers.ClusterDrivers.List(),
+		EnabledDriver:    r.Context.GatewayConfig().Spec.Cortex.Management.ClusterDriver,
+	}
+}
+
+func (r *RBACBackendService) Info(_ context.Context, capability *corev1.Reference) (*capabilityv1.Details, error) {
+	// Info must not block
+	if capability.GetId() != wellknown.CapabilityMetrics {
+		return nil, status.Errorf(codes.InvalidArgument, "capability %s not implemented by this plugin", capability.GetId())
+	}
+	return r.info(), nil
+}
+
+func (r *RBACBackendService) List(_ context.Context, _ *emptypb.Empty) (*capabilityv1.DetailsList, error) {
+	return &capabilityv1.DetailsList{
+		Items: []*capabilityv1.Details{
+			r.info(),
+		},
+	}, nil
+}
+
+func (r *RBACBackendService) GetAvailablePermissions(_ context.Context, _ *emptypb.Empty) (*corev1.AvailablePermissions, error) {
+	return &corev1.AvailablePermissions{
+		Items: []*corev1.PermissionDescription{
+			{
+				Type: string(corev1.PermissionTypeCluster),
+				Verbs: []*corev1.PermissionVerb{
+					corev1.VerbGet(),
+				},
+			},
+		},
+	}, nil
+}
+
+func (r *RBACBackendService) GetRole(ctx context.Context, in *corev1.Reference) (*corev1.Role, error) {
+	var revision int64
+	role, err := r.rolesStore.Get(ctx, in.GetId(), storage.WithRevisionOut(&revision))
+	if err != nil {
+		return nil, err
+	}
+	role.Metadata = &corev1.RoleMetadata{
+		ResourceVersion: strconv.FormatInt(revision, 10),
+	}
+	return role, nil
+}
+
+func (r *RBACBackendService) CreateRole(ctx context.Context, in *corev1.Role) (*emptypb.Empty, error) {
+	if err := validation.Validate(in); err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	_, err := r.rolesStore.Get(ctx, in.Reference().GetId())
+	if err == nil {
+		return nil, storage.ErrAlreadyExists
+	}
+	if !storage.IsNotFound(err) {
+		return nil, err
+	}
+
+	locker := system.NewLock(ctx, r.Context.KeyValueStoreClient(), lockKey(in.Reference().GetId()))
+
+	var innerErr error
+	err = locker.Try(func() {
+		innerErr = r.rolesStore.Put(ctx, in.GetId(), in)
+	}, func() {
+		innerErr = status.Error(codes.Aborted, "could not obtain lock")
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &emptypb.Empty{}, innerErr
+}
+
+func (r *RBACBackendService) UpdateRole(ctx context.Context, in *corev1.Role) (*emptypb.Empty, error) {
+	if err := validation.Validate(in); err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	locker := system.NewLock(ctx, r.Context.KeyValueStoreClient(), lockKey(in.Reference().GetId()))
+	var innerErr error
+
+	err := locker.Do(func() {
+		oldRole, err := r.rolesStore.Get(ctx, in.Reference().GetId())
+		if err != nil {
+			innerErr = err
+			return
+		}
+		revision, err := strconv.ParseInt(in.GetMetadata().GetResourceVersion(), 10, 64)
+		if err != nil {
+			innerErr = err
+			return
+		}
+		oldRole.Permissions = in.GetPermissions()
+		innerErr = r.rolesStore.Put(ctx, oldRole.Reference().GetId(), oldRole, storage.WithRevision(revision))
```

Comment:
if the role has immutable fields, we could consider defining a helper `apply(old, new)` function or `cmp(old, new)` error function, to prevent improper code changes down the line

Or, perhaps a change to the `rolestore` interface with an `Update` method that encapsulates the immutable field logic if it is shared between implementations

Reply:
This is a direct conversion of the existing logic.  I think improvements like this could be made separately.